### ISSUE TITLE
Namespace UI Components

### ISSUE THREAD

**Issue by SethDavenport** [rangle/rangle-starter#52]
The starter app breaks down UI into granular UI components, consistent with our practice.  However they are not namespaced to the project, which is inconsistent with our practice.

This leads to the risk of name clashes with other libraries, and in some cases with HTML itself (which is bad).

**Comment by SethDavenport** [comment]
So unless anyone screams foul I suggest we adopt the convention of a project-specific prefix for angular1/2 selectors, component class names, and JSX components. However, we don't need to bother prefixing the actual file names, because of bundling.

So for example, if the starters adopt the prefix 'rio' (rangle.io), we would have the following examples:

**ng2**

`components/form.ts`:

``` typescript
@Component({
  selector: 'rio-form',
  // ...
})
export class RioForm {
}
```

**ng1**

`components/form.ts`:

``` typescript
export class RioForm {
  selector: 'rio-form',
  ddo: {
     // ...
  }
}
```

**react**

`components/form.js`

``` JSX
export RioForm({prop1, prop2}) {
  return (<form>...</form>);
}
```

**Comment by SethDavenport** [comment]
@neilff @e-schultz @bennett000 @bertrandk

**Comment by neilff** [comment]
Looks fine to me. I understand the angular change, not sure the React one will provide as much value as I am still able to rename my component when I import it. e.g. `import RioFormRenamed from './form';` (this assumes it is a `default export`).

**Comment by kosz** [comment]
speaking of `export default` maybe we need to decide on a standard for exporting _components_ ?

React Components ( just like Angular Components ) normally have a dedicated file. 

does it make sense to use `export default` all the time or `export` ? Because that could impact this issue. If we use `export default` then we don't need to give them special names, we just need to import them with special names. 

At that point do we still need prefixing like in Angular ? 

Would this concern be as pressing as in Angular 1 where all the angular html tags were a global thing, vs this situation where we have the `import` and `JSX` buffers ? 

```
import {FormComponent} from 'external-form-component-lib';
import MyOwnRenamedFormBecauseIImportTheOtherFormHereFromTheOtherLib from './form';

...
<MyOwnRenamedFormBecauseIImportTheOtherFormHereFromTheOtherLib>
   Unlike Angular 1 without `replace`, This gets rendered as a div anyways,
    so would it ever conflict with HTML ?
</MyOwnRenamedFormBecauseIImportTheOtherFormHereFromTheOtherLib>
```

**Comment by SethDavenport** [comment]
Yeah in React it seems more common to export default, since you're doing one component per file.

In Angular, non-default exports seem to be the norm.  You should still be doing more than one export per file, but the community seems to non-default things.  Maybe because typescript occasionally means  a component file exports a component and a few supporting interfaces?

Nonetheless, prefixing is definitely needed on selectors, in both Angular 1 and 2.

**Comment by SethDavenport** [comment]
In angular code, I've also started seeing a log of `index.ts` files that lump together exports from subfolders.  I'm not sure how I feel about this to be honest.  Feels a bit redundant.

Maybe this is a topic for the next architecture guild.

**Comment by kosz** [comment]
I feel that index.ts aggregating pattern could be good for a framework, where you don't want your end users to remember many paths and file names. 

```
import { This, That } from 'angular/core';
```

However I'm thinking this may be not be as useful for an actual application ( as opposed to a framework ).

Actually let me rephrase that : 

Maybe this is a good pattern to use for highly reusable almost framework level components, but maybe not so good for very specific, business oriented components.

**Comment by SethDavenport** [comment]
Yeah - it's nice for importing stuff from angular2 for example.

However for UI components I find it a bit onerous.  Even if you have a folder for each component (due to external CSS or whatever), I prefer a patter more like:

```
src/components/
  MyComponent1/index.ts
                            my-component-1.css
  MyComponent2/index.ts
                            my-component-2.css
```

... etc.

This way you can still import a component using

```
import { MyComponent1 } from './components/MyComponent1';
```

I'm not dogmatic about it though.

**Comment by SethDavenport** [comment]
@neilff One question about prefixing components, even if they are renamable default exports, is whether it affects the name that shows up in react dev tools.

Is it inferred from the thing being exported or the thing being imported?  If it's the former we could still have name collisions in the dev tool.

E.g.

``` JavaScript
export default function MyComponent() {
 ...
};
```

``` JavaScript
import Whatever from './my-component'

function Container() {
  return <Whatever />;
};
```

Will it show up in dev tools as `MyComponent` or `Whatever`?

**Comment by SethDavenport** [comment]
This should wait until after #56 is done.

**Comment by SethDavenport** [comment]
Alan and Matt have been namespacing ng1 as they proceed with #56.  I have namespaced ng2.

Per internal discussions, we'll leave the react projects un-namespaced, since it's not necessary for react.
